### ISSUE TITLE
gpuOwl Fails with Memory Access Fault Error

### ISSUE THREAD

**Issue by valeriob01** [RadeonOpenCompute/ROCm#1020]
"Issue: gpuOwL is an OpenCL-based program for testing Mersenne numbers for primality. Currently, running gpuOwl for higher probable prime (PRP) values results in a Memory Access Fault error.

Note, the issue is noticed only when using higher PRP values.

Resolution: As a workaround, you may use lower PRP values."

This Resolution is not valid as PRP values are growing.

**Comment by Moading** [comment]
Hi,
have you tried to track down the problem with rocr_debug_agent? (https://github.com/ROCm-Developer-Tools/rocr_debug_agent)
I had a memory access fault in my application that I tried to find for months without luck. rocr_debug_agent pointed me to the right location in the code.
You need a Vega GPU, older ASICS are not supported.

**Comment by valeriob01** [comment]
> Hi,
> have you tried to track down the problem with rocr_debug_agent? (https://github.com/ROCm-Developer-Tools/rocr_debug_agent)
> I had a memory access fault in my application that I tried to find for months without luck. rocr_debug_agent pointed me to the right location in the code.
> You need a Vega GPU, older ASICS are not supported.

and Radeon VII is supported?

**Comment by Moading** [comment]
Yes

**Comment by preda** [comment]
At this point I have no indication that there is a memory access error in GpuOwl. Secondly, I am not able to run OpenCL with ROCm 3.0 as discussed here https://github.com/RadeonOpenCompute/ROCm/issues/977 , and as such I can't investigate GpuOwl's behavior on 3.0.

**Comment by valeriob01** [comment]
> At this point I have no indication that there is a memory access error in GpuOwl. Secondly, I am not able to run OpenCL with ROCm 3.0 as discussed here #977 , and as such I can't investigate GpuOwl's behavior on 3.0.

I would be useful if they indicated which PRP values they used for the test.

**Comment by preda** [comment]
As mentioned in #977 recently I was able to run with ROCm 3.1. There was no "memory access fault" reported but the computation does not proceed correctly. Basically there is a bug, either in gpuowl or in the OpenCL codegen or somewhere.

**Comment by pramenku** [comment]
> 
> 
> As mentioned in #977 recently I was able to run with ROCm 3.1. There was no "memory access fault" reported but the computation does not proceed correctly. Basically there is a bug, either in gpuowl or in the OpenCL codegen or somewhere.

it looks like it may be from application side on ROCm3.1.
We  can give a try below info and see if it helps.

gpuowl : https://github.com/preda/gpuowl 
Commit ID à  305d2b4949b7827af6e394eaf6c7a0d41c1b17c3  --> Pass
Commit ID à  6d275b7130dbcc05d9ca1771fd8ba2522f3b6a75 -->Fail [ first commit that started seeign failure ]

**Comment by preda** [comment]
> it looks like it may be from application side on ROCm3.1.
> We can give a try below info and see if it helps.
> 
> gpuowl : https://github.com/preda/gpuowl
> Commit ID à 305d2b4949b7827af6e394eaf6c7a0d41c1b17c3 --> Pass
> Commit ID à 6d275b7130dbcc05d9ca1771fd8ba2522f3b6a75 -->Fail [ first commit that started seeign failure ]

@pramenku Thank you I'll have a look.

**Comment by preda** [comment]
If that's fine I'm going to close this issue, because it seems there is no Memory Acccess Error with ROCm 3.1 anymore, so this issue relative to 3.0 seems to have been already fixed in 3.1.

**Comment by pskumar152** [comment]
No Memory Access Fault issue is observed in ROCm 3.1. 

Verified with the latest gpuowl commit + ROCm 3.1.

**Observations**
kumar@desktop-radeon7:~/Desktop/gpuowl$ ./gpuowl -prp 84682337
2020-03-02 12:57:06 gpuowl v6.11-160-g74b7196-dirty
2020-03-02 12:57:06 Note: not found 'config.txt'
2020-03-02 12:57:06 config: -prp 84682337
2020-03-02 12:57:06 device 0, unique id 'a5de692172dc768b'
2020-03-02 12:57:06 a5de692172dc768b 84682337 FFT 4608K: Width 256x4, Height 64x4, Middle 9; 17.95 bits/word
2020-03-02 12:57:07 a5de692172dc768b OpenCL args "-DEXP=84682337u -DWIDTH=1024u -DSMALL_HEIGHT=256u -DMIDDLE=9u -DWEIGHT_STEP=0x1.09aaaa1a94ac2p+0 -DIWEIGHT_STEP=0x1.ed5ec2690d837p-1 -DWEIGHT_BIGSTEP=0x1.ae89f995ad3adp+0 -DIWEIGHT_BIGSTEP=0x1.306fe0a31b715p-1 -DPM1=0 -DAMDGPU=1 -cl-fast-relaxed-math -cl-std=CL2.0"
2020-03-02 12:57:10 a5de692172dc768b OpenCL compilation in 3.40 s
2020-03-02 12:57:11 a5de692172dc768b 84682337 OK 0 loaded: blockSize 400, 0000000000000003
2020-03-02 12:57:12 a5de692172dc768b 84682337 OK 800 0.00%; 595 us/it; ETA 0d 14:00; d0b20c358eaddcea (check 0.33s)
2020-03-02 12:59:12 a5de692172dc768b 84682337 OK 200000 0.24%; 601 us/it; ETA 0d 14:07; 3ad53bdf0f0dcb10 (check 0.33s)
2020-03-02 13:01:13 a5de692172dc768b 84682337 OK 400000 0.47%; 603 us/it; ETA 0d 14:07; 2fa7256aa0425861 (check 0.33s)

**Comment by seesturm** [comment]
I was courious what actually fixed the memory access fault in ROCm 3.1. Didn't find any message from the maintainers on this topic. It appears that the fix is in the firmware blob since I had to update the firmware to the latest one found on linux-firmware.git.

Really a pity that there is no proper changelog for the firmware. One finding from this is that there are certain kind of bugs which cannot be fixed by the open source community despite all the host software being open source. But to be honest, currently there seems to be practically no community outside of AMD for maintaining ROCm.

**Comment by preda** [comment]
I still see a strange crash manifested in dmesg such:
```
[  501.631399] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631406] amdgpu 0000:67:00.0:   in page starting at address 0x00006f20054e4000 from client 27
[  501.631410] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631412] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631415] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631417] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631419] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631421] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631430] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631434] Evicting PASID 0x8003 queues
[  501.631435] amdgpu 0000:67:00.0:   in page starting at address 0x00005f610dc33000 from client 27
[  501.631439] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631441] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631444] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631445] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631447] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631449] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631454] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631457] amdgpu 0000:67:00.0:   in page starting at address 0x00005d6107692000 from client 27
[  501.631460] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631462] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631464] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631466] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631468] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631470] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631475] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631478] amdgpu 0000:67:00.0:   in page starting at address 0x00007f20054e4000 from client 27
[  501.631480] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631482] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631484] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631486] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631489] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631491] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631496] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631498] amdgpu 0000:67:00.0:   in page starting at address 0x00007f2107552000 from client 27
[  501.631501] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631503] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631505] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631507] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631509] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631511] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631516] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631519] amdgpu 0000:67:00.0:   in page starting at address 0x00005d6005426000 from client 27
[  501.631521] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631523] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631525] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631527] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631529] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631531] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631536] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631539] amdgpu 0000:67:00.0:   in page starting at address 0x00005f6105426000 from client 27
[  501.631541] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631543] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631545] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631547] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631550] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631552] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631557] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631559] amdgpu 0000:67:00.0:   in page starting at address 0x00007d6107692000 from client 27
[  501.631562] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631564] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631566] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631568] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631570] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631572] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631577] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631580] amdgpu 0000:67:00.0:   in page starting at address 0x00007f21075f2000 from client 27
[  501.631582] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631584] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631586] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631588] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631590] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631592] amdgpu 0000:67:00.0: 	 RW: 0x1
[  501.631597] amdgpu 0000:67:00.0: [gfxhub0] retry page fault (src_id:0 ring:0 vmid:8 pasid:32771, for process gpuowl pid 4376 thread gpuowl pid 4376)
[  501.631600] amdgpu 0000:67:00.0:   in page starting at address 0x00007c2109991000 from client 27
[  501.631602] amdgpu 0000:67:00.0: VM_L2_PROTECTION_FAULT_STATUS:0xFFFFFFFF
[  501.631604] amdgpu 0000:67:00.0: 	 MORE_FAULTS: 0x1
[  501.631606] amdgpu 0000:67:00.0: 	 WALKER_ERROR: 0x7
[  501.631608] amdgpu 0000:67:00.0: 	 PERMISSION_FAULTS: 0xf
[  501.631610] amdgpu 0000:67:00.0: 	 MAPPING_ERROR: 0x1
[  501.631612] amdgpu 0000:67:00.0: 	 RW: 0x1
[  510.633583] qcm fence wait loop timeout expired
[  510.633586] The cp might be in an unrecoverable state due to an unsuccessful queues preemption
[  510.633592] amdgpu 0000:67:00.0: GPU reset begin!
[  510.633595] Evicting PASID 0x8003 queues
[  510.633596] Evicting PASID 0x8003 queues
[  511.030771] amdgpu: [powerplay] Failed to send message 0x7, response 0xffffffff
[  511.030774] amdgpu: [powerplay] [DisableAllSMUFeatures] Failed to disable all smu features!
[  511.030774] amdgpu: [powerplay] [DisableDpmTasks] Failed to disable all smu features!
[  511.030775] amdgpu: [powerplay] [PowerOffAsic] Failed to disable DPM!
[  511.030836] [drm:amdgpu_device_ip_suspend_phase2 [amdgpu]] *ERROR* suspend of IP block <powerplay> failed -5
[  511.271180] [drm] Timeout wait for RLC serdes 0,0
[  511.292052] amdgpu 0000:67:00.0: GPU BACO reset
[  511.292055] amdgpu: [powerplay] Failed to send message 0x58, response 0xffffffff
[  511.292058] amdgpu 0000:67:00.0: GPU reset succeeded, trying to resume
[  511.292145] [drm] PCIE GART of 512M enabled (table at 0x0000008000000000).
[  511.292157] [drm] VRAM is lost due to GPU reset!
[  511.292273] [drm] PSP is resuming...
[  511.292277] amdgpu 0000:67:00.0: sos fw version = 0xffffffff.
[  511.431761] [drm:psp_hw_start [amdgpu]] *ERROR* PSP create ring failed!
[  511.431814] [drm:psp_resume [amdgpu]] *ERROR* PSP resume failed
[  511.431846] [drm:amdgpu_device_fw_loading [amdgpu]] *ERROR* resume of IP block <psp> failed -62
[  511.431878] [drm] Skip scheduling IBs!
[  511.431882] [drm] Skip scheduling IBs!
[  511.431885] [drm] Skip scheduling IBs!
[  511.431888] [drm] Skip scheduling IBs!
[  511.431891] [drm] Skip scheduling IBs!
[  511.431893] amdgpu 0000:67:00.0: GPU reset(1) failed
[  511.431894] [drm] Skip scheduling IBs!
[  511.431896] [drm] Skip scheduling IBs!
[  511.431899] [drm] Skip scheduling IBs!
[  511.431902] [drm] Skip scheduling IBs!
[  511.431904] [drm] Skip scheduling IBs!
[  511.431907] [drm] Skip scheduling IBs!
[  511.431909] [drm] Skip scheduling IBs!
[  511.431912] [drm] Skip scheduling IBs!
[  511.431915] [drm] Skip scheduling IBs!
```
The app reports
```
Memory access fault by GPU node-2 (Agent handle: 0x5561bed44740) on address 0x7d6107692000. Reason: Unknown.
```

This is with ROCm 3.1 on linux kernel 5.6.0-rc1. This system worked fine for a few days, I'm still investigating what changed that might have activated this issue. Anyway, if ROCm devs have any insight into what may cause this, or workarounds, I'd like to hear. It seems the problem does not originate application-side (i.e. it's not caused by a genuine invalid memory access).

**Comment by pramenku** [comment]
> 
> 
> If that's fine I'm going to close this issue, because it seems there is no Memory Acccess Error with ROCm 3.1 anymore, so this issue relative to 3.0 seems to have been already fixed in 3.1.
> (can't close, I think original author should close)

Thanks for the confirmation.

**Comment by streamhsa** [comment]
As per preda, 3.1 doesn't exhibit this issue. so closing this issue.
